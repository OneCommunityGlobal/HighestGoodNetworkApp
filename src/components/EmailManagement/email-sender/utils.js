import { Validators } from './validation';

export function parseRecipients(recipientText) {
  if (!recipientText || typeof recipientText !== 'string') return [];
  return recipientText
    .split(/[,;\n]/)
    .map(email => email.trim())
    .filter(email => email.length > 0 && email.includes('@'));
}

export function validateEmail(email) {
  if (!email || typeof email !== 'string') return false;
  const emailRegex = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
  return emailRegex.test(email.trim());
}

export function buildRenderedEmailFromTemplate(templateData, variableValues) {
  if (!templateData) return { subject: '', content: '' };

  let content = templateData.html_content || templateData.content || '';
  let subject = templateData.subject || '';

  if (Array.isArray(templateData.variables)) {
    templateData.variables.forEach(variable => {
      if (!variable || !variable.name) return;
      // Keep {{variableName}} as is if no value provided, don't replace with [variableName]
      if (!variableValues?.[variable.name]) return;

      let value = variableValues[variable.name];
      if (variable.type === 'image') {
        const extracted = variableValues?.[`${variable.name}_extracted`];
        if (extracted) value = extracted;
        else if (value) {
          const candidate = Validators.extractImageFromSource(value);
          if (candidate) value = candidate;
        }
      }
      const regex = new RegExp(`{{${variable.name}}}`, 'g');
      content = content.replace(regex, value);
      subject = subject.replace(regex, value);
    });
  }

  return { subject, content };
}
