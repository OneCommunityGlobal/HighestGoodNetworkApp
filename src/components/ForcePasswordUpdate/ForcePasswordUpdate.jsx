import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';
import Joi from 'joi';
import { toast } from 'react-toastify';
import { forcePasswordUpdate } from '../../actions/updatePassword';
import { clearErrors } from '../../actions/errorsActions';
import Form from '../common/Form';

export class ForcePasswordUpdate extends Form {
  state = {
    data: { newpassword: '', confirmnewpassword: '' },
    errors: {},
  };

  componentDidMount() {
    // document.title = "Force Update Password";
  }

  componentDidUpdate(prevProps) {
    if (prevProps.errors.error !== this.props.errors.error) {
      this.setState({ errors: this.props.errors });
    }
  }

  componentWillUnmount() {
    this.props.clearErrors();
  }

  schema = {
    newpassword: Joi.string()
      .regex(/(?=^.{8,}$)((?=.*\d)|(?=.*\W+))(?![.\n])(?=.*[A-Z])(?=.*[a-z]).*$/)
      .required()
      .label('New Password')
      .options({
        language: {
          string: {
            regex: {
              base:
                'should be at least 8 characters long and must include at least one uppercase letter, one lowercase letter, and one number or special character',
            },
          },
        },
      }),

    confirmnewpassword: Joi.any()
      .valid(Joi.ref('newpassword'))
      .options({ language: { any: { allowOnly: 'must match new password' } } })
      .label('Confirm Password'),
  };

  doSubmit = async () => {
    const { newpassword } = {
      ...this.state.data,
    };

    const { userId } = this.props.match.params;
    const data = { userId, newpassword };
    const status = await this.props.forcePasswordUpdate(data);
    if (status === 200) {
      toast.success(
        'You will now be directed to the login page where you can login with your new password.',
        {
          onClose: () => this.props.history.replace('/login'),
        },
      );
    } else {
      toast.error('Something went wrong. Please contact your administrator.');
    }
  };

  render() {
    const { darkMode } = this.props;

    return (
      <div
        className={`pt-5 h-100 container-fluid d-flex flex-column align-items-center ${
          darkMode ? 'bg-oxford-blue' : ''
        }`}
      >
        <h2>Change Password</h2>

        <form className="col-md-4 xs-12" onSubmit={e => this.handleSubmit(e)}>
          {this.renderInput({
            name: 'newpassword',
            label: 'New Password:',
            type: 'password',
            darkMode,
          })}
          {this.renderInput({
            name: 'confirmnewpassword',
            label: 'Confirm Password:',
            type: 'password',
            'data-refers': 'newpassword',
            darkMode,
          })}
          {this.renderButton({ label: 'Submit', darkMode })}
        </form>
      </div>
    );
  }
}

const mapStateToProps = state => ({
  errors: state.errors,
  darkMode: state.theme.darkMode,
});

export default withRouter(
  connect(mapStateToProps, {
    forcePasswordUpdate,
    clearErrors,
  })(ForcePasswordUpdate),
);
