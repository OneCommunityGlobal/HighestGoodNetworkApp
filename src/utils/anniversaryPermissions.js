export const calculateAnniversaryDate = (createdDate) => {
    let dt = new Date(createdDate)
    const sixMonthAnniversary = dt.setMonth(dt.getMonth() + 6)
    const oneWeekAfter6M = new Date(sixMonthAnniversary).getTime() + 7 * 24 * 60 * 60 * 1000 + 1
  
    dt = new Date(createdDate)
    const oneYearAnniversary = dt.setFullYear(dt.getFullYear() + 1)
    const oneWeekAfter1Y = new Date(oneYearAnniversary).getTime() + 7 * 24 * 60 * 60 * 1000 + 1
  
    dt = new Date(createdDate)
    const twoYearAnniversary = dt.setFullYear(dt.getFullYear() + 2)
    const oneWeekAfter2Y = new Date(twoYearAnniversary).getTime() + 7 * 24 * 60 * 60 * 1000 + 1
  
    dt = new Date(createdDate)
    const threeYearAnniversary = dt.setFullYear(dt.getFullYear() + 3)
    const oneWeekAfter3Y = new Date(threeYearAnniversary).getTime() + 7 * 24 * 60 * 60 * 1000 + 1
  
    dt = new Date(createdDate)
    const fourYearAnniversary = dt.setFullYear(dt.getFullYear() + 4)
    const oneWeekAfter4Y = new Date(fourYearAnniversary).getTime() + 7 * 24 * 60 * 60 * 1000 + 1
  
    dt = new Date(createdDate)
    const fiveYearAnniversary = dt.setFullYear(dt.getFullYear() + 5)
    const oneWeekAfter5Y = new Date(fiveYearAnniversary).getTime() + 7 * 24 * 60 * 60 * 1000 + 1
  
    dt = new Date(createdDate)
    const sixYearAnniversary = dt.setFullYear(dt.getFullYear() + 6)
    const oneWeekAfter6Y = new Date(sixYearAnniversary).getTime() + 7 * 24 * 60 * 60 * 1000 + 1
  
    dt = new Date(createdDate)
    const sevenYearAnniversary = dt.setFullYear(dt.getFullYear() + 7)
    const oneWeekAfter7Y = new Date(sevenYearAnniversary).getTime() + 7 * 24 * 60 * 60 * 1000 + 1
  
    dt = new Date(createdDate)
    const eightYearAnniversary = dt.setFullYear(dt.getFullYear() + 8)
    const oneWeekAfter8Y = new Date(eightYearAnniversary).getTime() + 7 * 24 * 60 * 60 * 1000 + 1
  
    dt = new Date(createdDate)
    const nineYearAnniversary = dt.setFullYear(dt.getFullYear() + 9)
    const oneWeekAfter9Y = new Date(nineYearAnniversary).getTime() + 7 * 24 * 60 * 60 * 1000 + 1
  
    dt = new Date(createdDate)
    const tenYearAnniversary = dt.setFullYear(dt.getFullYear() + 10)
    const oneWeekAfter10Y = new Date(tenYearAnniversary).getTime() + 7 * 24 * 60 * 60 * 1000 + 1
  
    let anniversaryDates = {
      sixMonthAnniversary: new Date(sixMonthAnniversary),
      oneWeekAfter6M: new Date(oneWeekAfter6M),
  
      oneYearAnniversary: new Date(oneYearAnniversary),
      oneWeekAfter1Y: new Date(oneWeekAfter1Y),
  
      twoYearAnniversary: new Date(twoYearAnniversary),
      oneWeekAfter2Y: new Date(oneWeekAfter2Y),
  
      threeYearAnniversary: new Date(threeYearAnniversary),
      oneWeekAfter3Y: new Date(oneWeekAfter3Y),
  
      fourYearAnniversary: new Date(fourYearAnniversary),
      oneWeekAfter4Y: new Date(oneWeekAfter4Y),
  
      fiveYearAnniversary: new Date(fiveYearAnniversary),
      oneWeekAfter5Y: new Date(oneWeekAfter5Y),
  
      sixYearAnniversary: new Date(sixYearAnniversary),
      oneWeekAfter6Y: new Date(oneWeekAfter6Y),
  
      sevenYearAnniversary: new Date(sevenYearAnniversary),
      oneWeekAfter7Y: new Date(oneWeekAfter7Y),
  
      eightYearAnniversary: new Date(eightYearAnniversary),
      oneWeekAfter8Y: new Date(oneWeekAfter8Y),
  
      nineYearAnniversary: new Date(nineYearAnniversary),
      oneWeekAfter9Y: new Date(oneWeekAfter9Y),
  
      tenYearAnniversary: new Date(tenYearAnniversary),
      oneWeekAfter10Y: new Date(oneWeekAfter10Y),
    };
  
    return anniversaryDates;
  }
  
  export const calculateDurationBetweenDates = (endDate, startDate) => {
  
    if (!startDate) {
      return { years: 0, months: 0, days: 0 };
    }

    let endDateObject = new Date(endDate)
    let startDateObject = new Date(startDate)
    let durationSinceStarted = {
      months: 0,
      years: 0
    }
    if (endDate > startDate) {
      var diff = Math.floor(endDateObject.getTime() - startDateObject.getTime());
      var day = 1000 * 60 * 60 * 24;
  
      var days = (diff / day);
      var months = (days / 31);
      var yearsSinceStarted = (months / 12);
      durationSinceStarted = {
        months: months,
        years: yearsSinceStarted
      }
    }
    else {
      durationSinceStarted = {
        months: 0,
        years: 0
      }
    }
    return durationSinceStarted
  }
  
  export const showTrophyIcon = (endDate, startDate) => {
    if (!startDate) {
      return false;
    }
  
    const calculateAnniversaryDateResults = calculateAnniversaryDate(startDate)
  
    if (startDate < endDate) {
      switch (true) {
        case calculateAnniversaryDateResults.tenYearAnniversary.toISOString().split('T')[0] <= endDate && calculateAnniversaryDateResults.oneWeekAfter10Y.toISOString().split('T')[0] > endDate:
          return true;
        case calculateAnniversaryDateResults.nineYearAnniversary.toISOString().split('T')[0] <= endDate && calculateAnniversaryDateResults.oneWeekAfter9Y.toISOString().split('T')[0] > endDate:
          return true;
        case calculateAnniversaryDateResults.eightYearAnniversary.toISOString().split('T')[0] <= endDate && calculateAnniversaryDateResults.oneWeekAfter8Y.toISOString().split('T')[0] > endDate:
          return true;
        case calculateAnniversaryDateResults.sevenYearAnniversary.toISOString().split('T')[0] <= endDate && calculateAnniversaryDateResults.oneWeekAfter7Y.toISOString().split('T')[0] > endDate:
          return true;
        case calculateAnniversaryDateResults.sixYearAnniversary.toISOString().split('T')[0] <= endDate && calculateAnniversaryDateResults.oneWeekAfter6Y.toISOString().split('T')[0] > endDate:
          return true;
        case calculateAnniversaryDateResults.fiveYearAnniversary.toISOString().split('T')[0] <= endDate && calculateAnniversaryDateResults.oneWeekAfter5Y.toISOString().split('T')[0] > endDate:
          return true;
        case calculateAnniversaryDateResults.fourYearAnniversary.toISOString().split('T')[0] <= endDate && calculateAnniversaryDateResults.oneWeekAfter4Y.toISOString().split('T')[0] > endDate:
          return true;
        case calculateAnniversaryDateResults.threeYearAnniversary.toISOString().split('T')[0] <= endDate && calculateAnniversaryDateResults.oneWeekAfter3Y.toISOString().split('T')[0] > endDate:
          return true;
        case calculateAnniversaryDateResults.twoYearAnniversary.toISOString().split('T')[0] <= endDate && calculateAnniversaryDateResults.oneWeekAfter2Y.toISOString().split('T')[0] > endDate:
          return true;
        case calculateAnniversaryDateResults.oneYearAnniversary.toISOString().split('T')[0] <= endDate && calculateAnniversaryDateResults.oneWeekAfter1Y.toISOString().split('T')[0] > endDate:
          return true;
        case calculateAnniversaryDateResults.sixMonthAnniversary.toISOString().split('T')[0] <= endDate && calculateAnniversaryDateResults.oneWeekAfter6M.toISOString().split('T')[0] > endDate:
          return true;
        default:
          false
      }
    }
    return false;
  }