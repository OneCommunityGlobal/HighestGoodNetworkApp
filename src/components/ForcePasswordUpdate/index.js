import ForcePasswordUpdate from './ForcePasswordUpdate';

export default ForcePasswordUpdate;
