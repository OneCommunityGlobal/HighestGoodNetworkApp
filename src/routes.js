import { lazy } from 'react';
import { Route, Switch } from 'react-router-dom';
import SetupProfile from 'components/SetupProfile/SetupProfile';
import { ToastContainer } from 'react-toastify';

// import SameFolderTasks from 'components/Projects/WBS/SameFolderTasks';
import AutoUpdate from 'components/AutoUpdate';
import { TaskEditSuggestions } from 'components/TaskEditSuggestions/TaskEditSuggestions';
import { RoutePermissions } from 'utils/routePermissions';
// import PermissionsManagement from 'components/PermissionsManagement/PermissionsManagement';
// import UserRoleTab from 'components/PermissionsManagement/UserRoleTab';
import EditableInfoModal from 'components/UserProfile/EditableModal/EditableInfoModal';
import RoleInfoCollections from 'components/UserProfile/EditableModal/roleInfoModal';
import LessonList from 'components/BMDashboard/LessonList/LessonListForm';
import AddEquipmentType from 'components/BMDashboard/Equipment/Add/AddEquipmentType';
import Announcements from 'components/Announcements';
import Timelog from './components/Timelog';
import LessonForm from './components/BMDashboard/Lesson/LessonForm';
// import Reports from './components/Reports';
// import UserProfile from './components/UserProfile';
import UserProfileEdit from './components/UserProfile/UserProfileEdit';
import Dashboard from './components/Dashboard';
import Logout from './components/Logout/Logout';
import Login from './components/Login';
import ForcePasswordUpdate from './components/ForcePasswordUpdate';
import ProtectedRoute from './components/common/ProtectedRoute';
import UpdatePassword from './components/UpdatePassword';
import Header from './components/Header';
// import Projects from './components/Projects';
// Teams from './components/Teams/Teams';
// import UserManagement from './components/UserManagement';
// import Members from './components/Projects/Members';
// import WBS from './components/Projects/WBS';
// import WBSDetail from './components/Projects/WBS/WBSDetail';
// import SingleTask from './components/Projects/WBS/SingleTask';
// import WeeklySummariesReport from './components/WeeklySummariesReport';
import TeamLocations from './components/TeamLocations';
import Admin from './components/Admin';
import 'react-toastify/dist/ReactToastify.css';
import { UserRole } from './utils/enums';
import ForgotPassword from './components/Login/ForgotPassword';
// import { PeopleReport } from './components/Reports/PeopleReport';
// import { ProjectReport } from './components/Reports/ProjectReport';
// import { TeamReport } from './components/Reports/TeamReport';
import Inventory from './components/Inventory';
// import BadgeManagement from './components/Badge/BadgeManagement';
import EmailSubscribeForm from './components/EmailSubscribeForm';
import UnsubscribeForm from './components/EmailSubscribeForm/Unsubscribe';



// BM Dashboard
import BMProtectedRoute from './components/common/BMDashboard/BMProtectedRoute';
import BMDashboard from './components/BMDashboard';
import BMLogin from './components/BMDashboard/Login';
import EquipmentList from './components/BMDashboard/Equipment/List';
import EquipmentDetail from './components/BMDashboard/Equipment/Detail/EquipmentDetail';
import ToolDetailPage from './components/BMDashboard/Tools/ToolDetailPage';
import CheckTypes from './components/BMDashboard/shared/CheckTypes';
import AddTool from './components/BMDashboard/Tools/AddTool';

const ReusableListView = lazy(() => import('./components/BMDashboard/ReusableList'));
const ConsumableListView = lazy(() => import('./components/BMDashboard/ConsumableList'));
const MaterialListView = lazy(() => import('./components/BMDashboard/MaterialList'));
const PurchaseMaterials = lazy(() => import('./components/BMDashboard/PurchaseRequests/MaterialPurchaseRequest'));
const PurchaseReusables = lazy(() => import('./components/BMDashboard/PurchaseRequests/ReusablePurchaseRequest'));
const ProjectDetails = lazy(() => import('./components/BMDashboard/Projects/ProjectDetails/ProjectDetails'));
const UpdateMaterialsBulk = lazy(() => import('./components/BMDashboard/UpdateMaterials/UpdateMaterialsBulk/UpdateMaterialsBulk'));
const InventoryTypesList = lazy(() => import('./components/BMDashboard/InventoryTypesList'));
const PurchaseTools = lazy(() => import('./components/BMDashboard/ToolPurchaseRequest'));
const AddMaterial = lazy(() => import('./components/BMDashboard/AddMaterial/AddMaterial'));
const AddConsumable = lazy(() => import('./components/BMDashboard/AddConsumable/AddConsumable'));
// Code-Splitting
const Projects = lazy(() => import('./components/Projects'));
const WeeklySummariesReport = lazy(() => import('./components/WeeklySummariesReport'));
const Reports = lazy(() => import('./components/Reports'));
const PeopleReport = lazy(() => import('./components/Reports/PeopleReport'));
const ProjectReport = lazy(() => import('./components/Reports/ProjectReport'));
const TeamReport = lazy(() => import('./components/Reports/TeamReport'));
const Members = lazy(() => import('./components/Projects/Members'));
const WBS = lazy(() => import('./components/Projects/WBS'));
const WBSDetail = lazy(() => import('./components/Projects/WBS/WBSDetail'));
const SingleTask = lazy(() => import('./components/Projects/WBS/SingleTask'));
const SameFolderTasks = lazy(() => import('./components/Projects/WBS/SameFolderTasks'));
const UserManagement = lazy(() => import('./components/UserManagement'));
const UserProfile = lazy(() => import('./components/UserProfile'));
const BadgeManagement = lazy(() => import('./components/Badge/BadgeManagement'));
const PermissionsManagement = lazy(() => import('./components/PermissionsManagement/PermissionsManagement'));
const UserRoleTab = lazy(() => import('./components/PermissionsManagement/UserRoleTab'));
const Teams = lazy(() => import('./components/Teams/Teams'));

export default (
  <Switch>
    <Route path="/ProfileInitialSetup/:token" component={SetupProfile} />
    <>
      {/* Comment out the Header component and its import during phase 2 development. */}
      <Header />
      {/* Uncomment BMHeader and its import during phase 2 development. */}
      {/* <BMHeader /> */}
      <AutoUpdate />
      <ToastContainer />
      <Switch>
        <ProtectedRoute path="/dashboard" exact component={Dashboard} />
        <ProtectedRoute path="/dashboard/:userId" exact component={Dashboard} />
        <ProtectedRoute path="/project/members/:projectId" fallback component={Members} />
        <ProtectedRoute path="/popupmanagement" component={Admin} />
        <ProtectedRoute path="/timelog/" exact component={Timelog} />
        <ProtectedRoute path="/timelog/:userId" exact component={Timelog} />
        <ProtectedRoute path="/peoplereport/:userId" component={PeopleReport} fallback />
        <ProtectedRoute path="/projectreport/:projectId" component={ProjectReport} fallback />
        <ProtectedRoute path="/teamreport/:teamId" component={TeamReport} fallback />
        <ProtectedRoute path="/taskeditsuggestions" component={TaskEditSuggestions} />

        <ProtectedRoute
          path="/inventory/:projectId"
          component={Inventory}
          routePermissions={RoutePermissions.inventoryProject}
        />
        <ProtectedRoute
          path="/inventory/:projectId/wbs/:wbsId"
          component={Inventory}
          routePermissions={RoutePermissions.inventoryProjectWbs}
        />

        <ProtectedRoute
          path="/weeklysummariesreport"
          exact
          component={WeeklySummariesReport}
          fallback
          allowedRoles={[
            UserRole.Administrator,
            UserRole.Manager,
            UserRole.CoreTeam,
            UserRole.Owner,
            UserRole.Mentor,
          ]}
          routePermissions={[RoutePermissions.weeklySummariesReport]}
        />
        <ProtectedRoute
          path="/reports"
          exact
          component={Reports}
          fallback
          routePermissions={RoutePermissions.reports}
        />
        <ProtectedRoute
          path="/teamlocations"
          exact
          component={TeamLocations}
        />
        <ProtectedRoute
          path="/projects"
          exact
          component={Projects}
          fallback
          allowedRoles={[UserRole.Administrator, UserRole.Owner, UserRole.Manager]}
          routePermissions={RoutePermissions.projects}
        />
        <ProtectedRoute
          path="/projects"
          exact
          component={Projects}
          fallback
          routePermissions={RoutePermissions.projects}
        />
        <ProtectedRoute path="/wbs/tasks/:wbsId/:projectId/:wbsName" component={WBSDetail} fallback />
        <ProtectedRoute path="/project/wbs/:projectId" component={WBS} fallback />
        <ProtectedRoute path="/wbs/tasks/:wbsId/:projectId" component={WBSDetail} fallback />
        <ProtectedRoute path="/wbs/tasks/:taskId" component={SingleTask} fallback />
        <ProtectedRoute path="/wbs/samefoldertasks/:taskId" component={SameFolderTasks} fallback />
        <ProtectedRoute
          path="/usermanagement"
          exact
          component={UserManagement}
          fallback
          routePermissions={RoutePermissions.userManagement}
        />
        <ProtectedRoute
          path="/badgemanagement"
          exact
          component={BadgeManagement}
          fallback
          routePermissions={RoutePermissions.badgeManagement}
        />
        <ProtectedRoute
          path="/permissionsmanagement"
          exact
          component={PermissionsManagement}
          fallback
          routePermissions={[
            RoutePermissions.permissionsManagement,
            RoutePermissions.userPermissionsManagement,
          ].flat()}
        />
        <ProtectedRoute
          path="/permissionsmanagement/:userRole"
          exact
          component={UserRoleTab}
          fallback
          routePermissions={RoutePermissions.permissionsManagement}
        />
        <ProtectedRoute
          path="/teams"
          exact
          component={Teams}
          fallback
          allowedRoles={[UserRole.Administrator, UserRole.Owner]}
          routePermissions={RoutePermissions.teams}
        />

        <ProtectedRoute
          path="/announcements"
          exact
          component={Announcements}
          allowedRoles={[UserRole.Administrator, UserRole.Owner]}
          routePermissions={RoutePermissions.projects}
        />

        {/* ----- BEGIN BM Dashboard Routing ----- */}
        <BMProtectedRoute path="/bmdashboard" exact component={BMDashboard} />
        <Route path="/bmdashboard/login" component={BMLogin} />

        <BMProtectedRoute path="/bmdashboard/materials/purchase" fallback component={PurchaseMaterials} />
        <BMProtectedRoute path="/bmdashboard/reusables/purchase" fallback component={PurchaseReusables} />
        <BMProtectedRoute path="/bmdashboard/tools/purchase" fallback component={PurchaseTools} />
        <BMProtectedRoute path="/bmdashboard/projects/:projectId" fallback component={ProjectDetails} />
        <BMProtectedRoute path="/bmdashboard/lessonlist/" component={LessonList} />
        <BMProtectedRoute path="/bmdashboard/materials/update" fallback component={UpdateMaterialsBulk} />
        <BMProtectedRoute path="/bmdashboard/materials/add" fallback component={AddMaterial} />
        <BMProtectedRoute path="/bmdashboard/equipment/add" component={AddEquipmentType} />
        <BMProtectedRoute path="/bmdashboard/inventory/types" component={CheckTypes} />
        <BMProtectedRoute path="/bmdashboard/equipment" fallback exact component={EquipmentList} />
        <BMProtectedRoute path="/bmdashboard/equipment/:equipmentId" component={EquipmentDetail} />
        <BMProtectedRoute path="/bmdashboard/consumables" component={ConsumableListView} />
        <BMProtectedRoute path="/bmdashboard/materials" fallback component={MaterialListView} />
        <BMProtectedRoute path="/bmdashboard/consumables/add" fallback component={AddConsumable} />
        <BMProtectedRoute path="/bmdashboard/consumables" fallback component={ConsumableListView} />
        <BMProtectedRoute path="/bmdashboard/reusables" fallback component={ReusableListView} />
        <BMProtectedRoute path="/bmdashboard/equipment/:equipmentId" fallback exact component={EquipmentDetail} />
        <BMProtectedRoute path="/bmdashboard/consumables" component={ConsumableListView} />
        <BMProtectedRoute path="/bmdashboard/equipment/:equipmentId" component={EquipmentDetail} />
        <BMProtectedRoute path="/bmdashboard/consumables" component={ConsumableListView} />
        <BMProtectedRoute path="/bmdashboard/tools/add" exact component={AddTool} />
        <BMProtectedRoute path="/bmdashboard/tools/:toolId" component={ToolDetailPage} />
        <BMProtectedRoute path="/bmdashboard/lessonform/:projectId" component={LessonForm} />
        <BMProtectedRoute path="/bmdashboard/lessonform/" component={LessonForm} />
        <BMProtectedRoute path="/bmdashboard/inventorytypes" fallback component={InventoryTypesList} />


        {/* Temporary route to redirect all subdirectories to login if unauthenticated */}
        <BMProtectedRoute path="/bmdashboard/:path" component={BMDashboard} />

        {/* ----- END BM Dashboard Routing ----- */}

        <Route path="/login" component={Login} />
        <Route path="/forgotpassword" component={ForgotPassword} />
        <Route path="/email-subscribe" component={EmailSubscribeForm} />
        <Route path="/email-unsubscribe" component={UnsubscribeForm} />
        <ProtectedRoute path="/infoCollections" component={EditableInfoModal} />
        <ProtectedRoute path="/infoCollections" component={RoleInfoCollections} />
        <ProtectedRoute path="/userprofile/:userId" fallback component={UserProfile} />
        <ProtectedRoute path="/userprofileedit/:userId" component={UserProfileEdit} />
        <ProtectedRoute path="/updatepassword/:userId" component={UpdatePassword} />
        <Route path="/Logout" component={Logout} />
        <Route path="/forcePasswordUpdate/:userId" component={ForcePasswordUpdate} />
        <ProtectedRoute path="/" exact component={Dashboard} />


      </Switch>
    </>
  </Switch>
);
