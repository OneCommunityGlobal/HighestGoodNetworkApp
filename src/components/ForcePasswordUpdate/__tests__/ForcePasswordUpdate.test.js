import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import { setupServer } from 'msw/node';
import { rest } from 'msw';
import { ENDPOINTS } from '../../../utils/URL';
import { ForcePasswordUpdate } from '../ForcePasswordUpdate';
import routes from '../../../routes';
import { renderWithRouterMatch } from '../../../__tests__/utils';
import mockState from '../../../__tests__/mockAdminState';
import '@testing-library/jest-dom/extend-expect';

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);
const mockClearErrors = jest.fn();

let passwordUpdated = false;

// Setup Mock Service Worker server
const server = setupServer(
  // PATCH request to update password
  rest.patch(ENDPOINTS.FORCE_PASSWORD, (req, res, ctx) => {
    passwordUpdated = true;
    return res(ctx.status(200), ctx.json({ message: 'Password updated successfully' }));
  }),

  // GET user profile
  rest.get('http://localhost:4500/api/userprofile/:userId', (req, res, ctx) => {
    return res(ctx.status(200), ctx.json({ id: req.params.userId, name: 'Test User' }));
  }),

  // GET timer service
  rest.get('http://localhost:4500/timer-service', (req, res, ctx) => {
    return res(ctx.status(200), ctx.json({ running: false }));
  }),

  // GET notifications unread
  rest.get('http://localhost:4500/api/notification/unread/user/:userId', (req, res, ctx) => {
    return res(ctx.status(200), ctx.json([])); // returning empty unread notifications
  }),

  // Mock dashboard fetch
  rest.get('http://localhost:4500/api/dashboard/*', (req, res, ctx) => {
    return res(ctx.status(200), ctx.json([]));
  }),

  // Catch all unmatched requests
  rest.get('*', (req, res, ctx) => {
    console.error(`Unhandled request: ${req.url.toString()}`);
    return res(ctx.status(500), ctx.json({ error: 'Unhandled request' }));
  }),
);

beforeAll(() => server.listen());
afterEach(() => {
  server.resetHandlers();
  passwordUpdated = false;
});
afterAll(() => server.close());

describe('ForcePasswordUpdate tests', () => {
  let store;
  const userId = '5edf141c78f1380017b829a6';

  beforeEach(() => {
    store = mockStore({
      auth: { user: { userid: userId }, isAuthenticated: true },
      theme: { darkMode: false },
      errors: {},
    });
  });

  it('should render the page and match snapshot', () => {
    const { asFragment } = render(
      <Provider store={store}>
        <ForcePasswordUpdate
          auth={{ isAuthenticated: true }}
          errors={{}}
          clearErrors={mockClearErrors}
          match={{ params: { userId } }}
        />
      </Provider>,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('should disable submit button if form is invalid', () => {
    render(
      <Provider store={store}>
        <ForcePasswordUpdate
          auth={{ isAuthenticated: true }}
          errors={{}}
          clearErrors={mockClearErrors}
          match={{ params: { userId } }}
        />
      </Provider>,
    );
    const submitButton = screen.getByRole('button', { name: /submit/i });
    expect(submitButton).toBeDisabled();
  });

  it('should show error if password requirements are not met', async () => {
    renderWithRouterMatch(routes, {
      initialState: mockState,
      route: `/forcePasswordUpdate/${userId}`,
    });

    fireEvent.change(screen.getByLabelText(/new password/i), { target: { value: 'short' } });
    await waitFor(() => {
      expect(screen.getByText(/at least 8 characters long/i)).toBeInTheDocument();
    });
  });

  it('should show error if passwords do not match', async () => {
    renderWithRouterMatch(routes, {
      initialState: mockState,
      route: `/forcePasswordUpdate/${userId}`,
    });

    fireEvent.change(screen.getByLabelText(/new password/i), { target: { value: 'NewPassword8' } });
    fireEvent.change(screen.getByLabelText(/confirm password/i), {
      target: { value: 'Mismatch8' },
    });

    fireEvent.click(screen.getByRole('button', { name: /submit/i }));

    await waitFor(() => {
      expect(screen.getByText(/must match new password/i)).toBeInTheDocument();
    });
  });

  it('should submit successfully if passwords match and requirements are met', async () => {
    renderWithRouterMatch(routes, {
      initialState: mockState,
      route: `/forcePasswordUpdate/${userId}`,
    });

    fireEvent.change(screen.getByLabelText(/new password/i), { target: { value: 'NewPassword8' } });
    fireEvent.change(screen.getByLabelText(/confirm password/i), {
      target: { value: 'NewPassword8' },
    });

    fireEvent.click(screen.getByRole('button', { name: /submit/i }));

    await waitFor(() => {
      expect(passwordUpdated).toBeTruthy();
      expect(screen.getByText(/you will now be directed to the login page/i)).toBeInTheDocument();
    });
  });
});
