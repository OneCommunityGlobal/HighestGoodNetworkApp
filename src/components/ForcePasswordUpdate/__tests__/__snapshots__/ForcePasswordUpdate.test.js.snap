// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Force Password Update page structure should match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="pt-5 h-100 container-fluid d-flex flex-column align-items-center "
  >
    <h2>
      Change Password
    </h2>
    <form
      class="col-md-4 xs-12"
    >
      <div
        class="form-group "
      >
        <label
          class=""
          for="newpassword"
        >
          New Password:
        </label>
        <div
          class="input-text w-100"
        >
          <input
            class="form-control"
            id="newpassword"
            name="newpassword"
            type="password"
            value=""
          />
          <i
            class="fa fa-eye-slash"
          />
        </div>
      </div>
      <div
        class="form-group "
      >
        <label
          class=""
          for="confirmnewpassword"
        >
          Confirm Password:
        </label>
        <div
          class="input-text w-100"
        >
          <input
            class="form-control"
            data-refers="newpassword"
            id="confirmnewpassword"
            name="confirmnewpassword"
            type="password"
            value=""
          />
          <i
            class="fa fa-eye-slash"
          />
        </div>
      </div>
      <button
        class="btn btn-primary"
        disabled=""
        style="box-shadow: 2px 2px 4px 1px lightgray;"
      >
        Submit
      </button>
    </form>
  </div>
</DocumentFragment>
`;